### PR TITLE
[onert] Introduce experimental training APIs

### PR DESCRIPTION
**Pull request by hseok-oh**
This commit introduces experimental training APIs.

ONE-DCO-1.0-Signed-off-by: Hyeongseok Oh <hseok82.oh@samsung.com>

### CODE REVIEW

**Comment by hseok-oh** [comment]
Two different style

- Add APIs on `nnfw_session` session
  - Split APIs as possible
- Introduce APIs on new `nnfw_train_session` session
  - Merge APIs as possible

File: runtime/onert/api/include/nnfw_experimental_train.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * @brief Get the number of training model inputs
+ *        This function should be called after {@link nnfw_prepare_train}
+ *
+ * @param[in]   session The session to get the number of training model inputs
+ * @param[out]  count   The number of training model inputs
+ * @return @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_get_input_count(nnfw_train_session *session, uint32_t *count);
```
Comment (glistening) [inline runtime/onert/api/include/nnfw_experimental_train.h:113]: @hseok-oh We may need use different name to distinguish the same name for inference (`nnfw_session` series).

**Comment by hseok-oh** [comment]
Updated

- Remove `nnfw_train_session`
- Split APIs to set input & expected
- Remove `nnfw_set_train_output`
- Add index to get loss for multiple output
- Add `update_weights` argument for validate mode

File: runtime/onert/api/include/nnfw_experimental.h [hunk 2]
```diff
@@ -152,4 +156,193 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+ * @note  This function should be called after {@link nnfw_prepare_train}
+ *
+ * @param session       The session to be set training inputs and expected model outputs
+ * @param index         The index of training expected output
+ * @param expected      The expected buffers for training
+ * @param expected_info The shape and type of expected buffer
+ *                      If it is nullptr, it will not change shape and batch size
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_set_train_expected(nnfw_session *session, uint32_t index, void *expected,
```
Comment (jyoungyun) [inline runtime/onert/api/include/nnfw_experimental.h:230]: When will this function be used? Is this function used with `nnfw_run` api ?

File: runtime/onert/api/include/nnfw_experimental.h [hunk 3]
```diff
@@ -152,4 +156,193 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+
+/**
+ * @brief Training information to prepare training
+ */
+typedef struct nnfw_train_info
+{
+  /** Learning rate */
+  float learning_rate = 0.001f;
+  /** Batch size */
+  uint32_t batch_size = 1;
```
Comment (jyoungyun) [inline runtime/onert/api/include/nnfw_experimental.h:187]: TODO
- loss
- optimizer

**Comment by hseok-oh** [comment]
@nnfw-bot test onert-x64-release

**Comment by glistening** [comment]
@hseok-oh If you determined not to introduce `nnfw_training_session`, what about using `nnfw_train` for training related APIs? For example,

- `nnfw_prepare_train` → `nnfw_tᅟrain_prepare`
- `nnfw_set_train_input` → `nnfw_train_set_input`
- `nnfw_set_train_expected` → `nnfw_train_set_expected`
- `nnfw_train` → `nnfw_train` ( ? )
- `nnfw_get_loss` → `nnfw_train_get_loss`
- `nnfw_export_inference_model` → `nnfw_train_export_model`
- `nnfw_get_train_input_count` → `nnfw_train_get_input_count`
...

**Comment by hseok-oh** [comment]
I've updated function name. PTAL

File: runtime/onert/api/include/nnfw_experimental.h [hunk 4]
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+/**
+ * @brief Export inference model
+ * @note  This function should be called on training mode
+ *        This function should be called after {@link nnfw_train}
+ *
+ * @param[in] session The session to export inference model
+ * @param[in] path    The path to export inference model
+ * @return @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_export_inference_model(nnfw_session *session, const char *path);
```
Comment (glistening) [inline runtime/onert/api/include/nnfw_experimental.h:266]: Will we export other type of model? I don't have an idea about other type of models except for inference model. Maybe we can export checkpoint later — `nnfw_train_export_checkpoint`.

File: runtime/onert/api/include/nnfw_experimental.h [hunk 5]
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+
+/**
+ * @brief Get the number of training model inputs
+ * @note  This function should be called after {@link nnfw_train_prepare}
+ *
+ * @param[in]   session The session to get the number of training model inputs
+ * @param[out]  count   The number of training model inputs
+ * @return @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_input_count(nnfw_session *session, uint32_t *count);
```
Comment (glistening) [inline runtime/onert/api/include/nnfw_experimental.h:280]: What is the difference between model input and train input? I guessed the count would be same.

File: runtime/onert/api/include/nnfw_experimental.h [hunk 6]
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+                                        nnfw_tensorinfo *info);
+
+/**
+ * @brief Get the number of training model expected outputs
+ * @note  This function should be called after {@link nnfw_train_prepare}
+ *
+ * @param[in]   session The session to get the number of training model expected outputs
+ * @param[out]  count   The number of training model expected outputs
+ */
+NNFW_STATUS nnfw_train_expected_count(nnfw_session *session, uint32_t *count);
```
Comment (glistening) [inline runtime/onert/api/include/nnfw_experimental.h:300]: Also, I guessed expected count == model output count. Why do we need this API?

File: runtime/onert/api/include/nnfw_experimental.h [hunk 7]
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+//////////////////////////////////////////////
+
+/**
+ * @brief Convert training model to inference model
+ * @note  This function should be called after {@link nnfw_train}
+ *
+ * @param session The session to convert training mode to inference mode
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_convert_to_inference(nnfw_session *session);
```
Comment (glistening) [inline runtime/onert/api/include/nnfw_experimental.h:324]: According to @param, it is for setting to training mode. 

```suggestion
NNFW_STATUS nnfw_train_set_training_mode(nnfw_session *session, bool is_training);
```

File: runtime/onert/api/include/nnfw_experimental.h [hunk 8]
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+
+/**
+ * @brief Convert inference model to training model
+ * @note  This function should be called after {@link nnfw_prepare}
+ *
+ * @param session The session to convert inference mode to training mode
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_convert_to_train(nnfw_session *session);
```
Comment (glistening) [inline runtime/onert/api/include/nnfw_experimental.h:333]: We may remove this API by `nnfw_train_set_training_mode(sess, false)`.

File: runtime/onert/api/include/nnfw_experimental.h [hunk 9]
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+
+/**
+ * @brief Training information to prepare training
+ * @todo  Add more training information
+ *        (e.g. optimizer, loss function, ...)
+ */
+typedef struct nnfw_train_info
+{
+  /** Learning rate */
+  float learning_rate = 0.001f;
```
Comment (jyoungyun) [inline runtime/onert/api/include/nnfw_experimental.h:186]: How about separating the train preparation info and execution info?
The `nnfw_train_prepare` does not need learning_rate information. This is needed in `nnfw_train`.

After the trained graph is generated once, the user can use this graph with various learning rate and epoch. Do we need a scenario like this?

Comment (hseok-oh) [reply runtime/onert/api/include/nnfw_experimental.h:186]: >How about separating the train preparation info and execution info?

I didn't consider this yet. But we may need this later. 
I supposed that we will get all parameter for training before `prepare` as our first step.

>After the trained graph is generated once, the user can use this graph with various learning rate

`nnfw_train_set_traininfo` API may be used for this.

File: runtime/onert/api/include/nnfw_experimental.h (hunk 6 above)
Comment (hseok-oh) [reply runtime/onert/api/include/nnfw_experimental.h:300]: Let's remove this and `nnfw_train_input_count`.

**Comment by hseok-oh** [comment]
I've updated
- Remove in/expected count APIs
- Rename export circle API
- Comment out non-essential APIs
- Fix typo

**Review by jyoungyun — DISMISSED** [review]
API design looks good to me. :)

I made a draft PR using this changes. #10900

File: runtime/onert/api/include/nnfw_experimental.h [hunk 10]
```diff
@@ -152,4 +156,168 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+ *
+ * @param[in] session     The session to be set training inputs and expected model outputs
+ * @param[in] index       The index of training input
+ * @param[in] input       The input buffers for training
+ * @param[in] input_info  The shape and type of input buffer
+ *                        If it is nullptr, it will not change shape and batch size
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, void *input,
+                                 nnfw_tensorinfo *input_info);
```
Comment (ragmani) [inline runtime/onert/api/include/nnfw_experimental.h:218]: Is it okay not to pass the size of buffer?

File: runtime/onert/api/include/nnfw_experimental.h [hunk 11]
```diff
@@ -152,4 +156,168 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+ * @note  This function should be called after {@link nnfw_train_prepare}
+ *
+ * @param[in] session     The session to be set training inputs and expected model outputs
+ * @param[in] index       The index of training input
+ * @param[in] input       The input buffers for training
+ * @param[in] input_info  The shape and type of input buffer
+ *                        If it is nullptr, it will not change shape and batch size
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, void *input,
```
Comment (ragmani) [inline runtime/onert/api/include/nnfw_experimental.h:217]:
```suggestion
NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, const void *input,
```

File: runtime/onert/api/src/nnfw_api.cc [hunk 12]
```diff
@@ -385,3 +385,53 @@ NNFW_STATUS nnfw_pop_pipeline_output(nnfw_session *session, void *outputs)
...
+
+// Training
+
+NNFW_STATUS nnfw_train_prepare(nnfw_session *session, nnfw_train_info *info)
+{
+  NNFW_RETURN_ERROR_IF_NULL(session);
+  return session->train_prepare(info);
+}
+
+NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, void *input,
```
Comment (ragmani) [inline runtime/onert/api/src/nnfw_api.cc:397]:
```suggestion
NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, const void *input,
```

File: runtime/onert/api/include/nnfw_experimental.h [hunk 13]
```diff
@@ -152,4 +156,168 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
...
+ * @note  This function should be called after {@link nnfw_train_prepare}
+ *
+ * @param session       The session to be set training inputs and expected model outputs
+ * @param index         The index of training expected output
+ * @param expected      The expected buffers for training
+ * @param expected_info The shape and type of expected buffer
+ *                      If it is nullptr, it will not change shape and batch size
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, void *expected,
```
Comment (ragmani) [inline runtime/onert/api/include/nnfw_experimental.h:231]:
```suggestion
NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, const void *expected,
```

File: runtime/onert/api/src/nnfw_api.cc [hunk 14]
```diff
@@ -385,3 +385,53 @@ NNFW_STATUS nnfw_pop_pipeline_output(nnfw_session *session, void *outputs)
...
+                                 nnfw_tensorinfo *input_info)
+{
+  NNFW_RETURN_ERROR_IF_NULL(session);
+  return session->train_set_input(index, input, input_info);
+}
+
+NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, void *expected,
```
Comment (ragmani) [inline runtime/onert/api/src/nnfw_api.cc:404]:
```suggestion
NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, const void *expected,
```

File: runtime/onert/api/src/nnfw_api_internal.cc [hunk 15]
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
 
   return NNFW_STATUS_NO_ERROR;
 }
+
+NNFW_STATUS nnfw_session::train_prepare(nnfw_train_info *info)
```
Comment (ragmani) [inline runtime/onert/api/src/nnfw_api_internal.cc:1133]:
```suggestion
NNFW_STATUS nnfw_session::train_prepare(const nnfw_train_info *info)
```

File: runtime/onert/api/src/nnfw_api_internal.cc [hunk 16]
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
...
+
+  // Use default info if info is null, otherwise use given info
+  (void)info;
+
+  // NYI
+  // _state = State::PREPARED_TRAINING;
+  return NNFW_STATUS_ERROR;
+}
+
+NNFW_STATUS nnfw_session::train_set_input(uint32_t index, void *input,
```
Comment (ragmani) [inline runtime/onert/api/src/nnfw_api_internal.cc:1155]:
```suggestion
NNFW_STATUS nnfw_session::train_set_input(uint32_t index, const void *input,
```

File: runtime/onert/api/src/nnfw_api_internal.cc [hunk 17]
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
...
+  // Maybe it cannot use general execute->setInput()
+  // because it needs to set batch input
+  (void)index;
+  (void)input_tensorinfo;
+
+  // NYI
+  return NNFW_STATUS_ERROR;
+}
+
+NNFW_STATUS nnfw_session::train_set_expected(uint32_t index, void *expected,
```
Comment (ragmani) [inline runtime/onert/api/src/nnfw_api_internal.cc:1181]:
```suggestion
NNFW_STATUS nnfw_session::train_set_expected(uint32_t index, const void *expected,
```

**Review by ragmani — COMMENTED** [review]
API design looks good to me. :)

File: runtime/onert/api/include/nnfw_experimental.h (hunk 10 above)
Comment (hseok-oh) [reply runtime/onert/api/include/nnfw_experimental.h:218]: `input_info` will be used to get size.

**Review by glistening — APPROVED** [review]
LGTM

**Review by jyoungyun — DISMISSED** [review]
LGTM

**Comment by hseok-oh** [comment]
I've updated to add APIs to get tensorinfo. PTAL

File: runtime/onert/api/src/nnfw_api_internal.h [hunk 18]
```diff
@@ -163,6 +163,8 @@ struct nnfw_session
   NNFW_STATUS set_backends_per_operation(const char *backend_settings);
 
   NNFW_STATUS train_prepare(const nnfw_train_info *info);
+  NNFW_STATUS train_input_tensorinfo(uint32_t index, nnfw_tensorinfo *ti);
+  NNFW_STATUS train_expected_tensorinfo(uint32_t index, nnfw_tensorinfo *ti);
```
Comment (jyoungyun) [inline runtime/onert/api/src/nnfw_api_internal.h:167]: 👍

File: runtime/onert/api/src/nnfw_api_internal.cc [hunk 19]
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
...
+    if (_state == State::PREPARED_TRAINING)
+      std::cerr << "prepare should be run once";
+    else
+      std::cerr << "invalid state";
+    std::cerr << std::endl;
+    return NNFW_STATUS_INVALID_STATE;
+  }
+
+  // Use default info if info is null, otherwise use given info
+  (void)info;
```
Comment (jyoungyun) [inline runtime/onert/api/src/nnfw_api_internal.cc:1148]: If the info is null, how about treating it as an error to clarify operation?

**Review by ragmani — APPROVED** [review]

**Review by jyoungyun — APPROVED** [review]
LGTM

**Review by glistening — APPROVED** [review]
LGTM

File: runtime/onert/api/src/nnfw_api_internal.cc (hunk 19 above)
Comment (hseok-oh) [reply runtime/onert/api/src/nnfw_api_internal.cc:1148]: It is designed to use training info in model metadata if `info` param is `nullptr`. Please refer API comment in `nnfw_experimental.h`.

File: runtime/onert/api/src/nnfw_api_internal.cc [hunk 20]
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
...
+
+NNFW_STATUS nnfw_session::train_set_input(uint32_t index, const void *input,
+                                          const nnfw_tensorinfo *input_tensorinfo)
+{
+  if (input == nullptr)
+  {
+    std::cerr << "Error during nnfw_session::train_set_input : input buffer is null" << std::endl;
+    return NNFW_STATUS_UNEXPECTED_NULL;
+  }
+
```
[the remark this answers is not in the document]
Comment (hseok-oh) [reply runtime/onert/api/src/nnfw_api_internal.cc:1196]: It is designed to use shape in model and batch training info if `input_tensorinfo` is `nullptr`. Please refer API comment in `nnfw_experimental.h`
